include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,621 @@
+#endif
+#ifdef DEAL_II_WITH_SUNDIALS
+      additional_data.solver_type = "kinsol";
+#endif
+
+      // If "auto" is still the solver type we cannot solve the problem
+      if (additional_data.solver_type == "auto")
+        AssertThrow(false, ExcMessage("No valid solver type."));
+    }
+
+  if (additional_data.solver_type == "kinsol")
+    {
+#ifdef DEAL_II_WITH_SUNDIALS
+      SUNDIALS::KINSOL<VectorType> nonlinear_solver(additional_data_kinsol);
+
+      nonlinear_solver.reinit_vector = [&](Vector<double> &x) {
+        NonlinearSolverSelector<VectorType>::reinit_vector(x);
+      };
+
+      nonlinear_solver.residual = [&](const Vector<double> &evaluation_point,
+                                      Vector<double> &      residual) {
+        return NonlinearSolverSelector<VectorType>::residual(evaluation_point,
+                                                             residual);
+      };
+
+      nonlinear_solver.setup_jacobian =
+        [&](const Vector<double> &current_u,
+            const Vector<double> /*&current_f*/) {
+          return NonlinearSolverSelector<VectorType>::setup_jacobian(current_u);
+        };
+
+      nonlinear_solver.solve_with_jacobian = [&](const Vector<double> &rhs,
+                                                 Vector<double> &      dst,
+                                                 const double tolerance) {
+        return NonlinearSolverSelector<VectorType>::solve_with_jacobian(
+          rhs, dst, tolerance);
+      };
+
+      nonlinear_solver.solve(initial_guess_and_solution);
+#else
+      AssertThrow(
+        false, ExcMessage("You do not have SUNDIALS configured with deal.II!"));
+#endif
+    }
+  else if (additional_data.solver_type == "nox")
+    {
+#ifdef DEAL_II_WITH_TRILINOS
+      TrilinosWrappers::NOXSolver<VectorType> nonlinear_solver(
+        additional_data_nox, parameters_nox);
+
+      nonlinear_solver.residual = [&](const Vector<double> &evaluation_point,
+                                      Vector<double> &      residual) {
+        return NonlinearSolverSelector<VectorType>::residual(evaluation_point,
+                                                             residual);
+      };
+
+      nonlinear_solver.setup_jacobian = [&](const Vector<double> &current_u) {
+        return NonlinearSolverSelector<VectorType>::setup_jacobian(current_u);
+      };
+
+      nonlinear_solver.solve_with_jacobian = [&](const Vector<double> &rhs,
+                                                 Vector<double> &      dst,
+                                                 const double tolerance) {
+        return NonlinearSolverSelector<VectorType>::solve_with_jacobian(
+          rhs, dst, tolerance);
+      };
+
+      nonlinear_solver.solve(initial_guess_and_solution);
+#else
+      AssertThrow(
+        false, ExcMessage("You do not have Trilinos configured with deal.II"));
+#endif
+    }
+  else
+    AssertThrow(false, ExcMessage("Invalid nonlinear solver specified"));
```

Comment:
Maybe say what the choice was, and what options are available given the current configuration?